data_model/src/predicate.rs

```diff
@@ -0,0 +1,50 @@
+use crate::query::QueryResult;
+
+pub trait Predicate{
+    type Input;
+
+    fn applies(&self, input: &Self::Input) -> bool;
+}
+
+pub struct DisplayContains(pub String);
+
+pub struct IdContains(pub String);
```

Comment:
What will be the format of the query then?
Maybe it consists of a target (e.g. `ValueType`) and predicates?

Reply:
The idea is that predicates are built up like legos. A predicate applies to a `Value`. It has an enum of targets: `metadata`, `id (Display)`, `self (Display)` and `self (Debug)` (which I'm not sure is a good idea to include). And then it also has an enum of operations: `fn (&str) -> bool` that map onto the string representations e.g. `is`, `contains`, and `not`. This should give you 12 filters at the cost of 7 with the added benefit of being easy to extend.

Reply:
We could then add combinators into the mix to have more fine-grained control.